impy/models/sophia.py

```diff
@@ -5,158 +5,191 @@
+        """Returns the array of indices of the decayed parent particles"""
+        MCEvent.parents(self)
+        return self.lib.schg.decpar[0:self.npart]
+
+    @property
+    def parents(self):
+        """In SOPHIA parents are difficult to obtain. This function returns 0."""
+        MCEvent.parents(self)
+        return self.lib.hepevt.jmohep[0:self.npart]
+
+    @property
+    def children(self):
+        """In SOPHIA daughters are difficult to obtain. This function returns 0."""
+        MCEvent.children(self)
+        return self.lib.hepevt.jdahep[0:self.npart]
+
+class SophiaRun(MCRun):
+    """Implements all abstract attributes of MCRun for the 
+    Sophia event generator.
+    """
+
+    def sigma_inel(self, *args, **kwargs):
+        """Inelastic cross section according to current
+        event setup (energy, projectile, target)"""
+        k = self._curr_event_kin
+        
+        if k.p1pdg != 22:
+            info(0, "The first particle must be a photon, but particle pdg = ", k.p1pdg)
+            raise Exception('Input error')
+        
+        if k.p2pdg not in [2212, 2112]:
+            info(0, "The second particle must be a proton or neutron, but particle pdg = ", k.p2pdg)
+            raise Exception('Input error')
+        
+        # self.energy_of_photon is the energy in lab frame
+        # where nucleon is at rest and photon is moving
+        total_crossection_id = 3 # 3 is for total crossection
+        # cross section in micro barn
+        return self.lib.crossection(self.energy_of_photon, total_crossection_id, 
+                                    self.nucleon_code_number)
+        
+
+    def sigma_inel_air(self):
+        """Inelastic cross section according to current
+        event setup (energy, projectile, target)"""
+        raise Exception('SophiaRun.sigma_inel_air has no implementation')
+
+    def set_event_kinematics(self, event_kinematics):
+        """Set new combination of energy, momentum, projectile
+        and target combination for next event."""
+
+        info(5, 'Setting event kinematics.')
+        info(10, event_kinematics)
+        k = event_kinematics
+        
+        if k.p1pdg != 22:
+            info(0, "The first particle must be a photon, but particle pdg = ", k.p1pdg)
+            raise Exception('Input error')
+        
+        if k.p2pdg not in [2212, 2112]:
+            info(0, "The second particle must be a proton or neutron, but particle pdg = ", k.p2pdg)
+            raise Exception('Input error')
+        
+        
+        self.nucleon_code_number = self.lib.icon_pdg_sib(k.p2pdg)
+        self.energy_of_nucleon = k.pmass2
+        self.energy_of_photon = k.elab 
+        # Here we consider laboratory frame where photon moves along z axis
+        # and nucleon is at rest. The angle is counted from z axis.
+        # However, because of the definitions in "eventgen" subroutine of
+        # SOPHIA code (line "P_gam(3) = -EPS*COS(theta*pi/180.D0)")
```

Comment:
Please, don't forget to add this as an issue after PR merged.

Reply:
I will add it